<a href="https://colab.research.google.com/github/Gilangp/Machine-learning/blob/main/week%2010/TG2_JS10_Gilang_Purnomo_2341720042.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Tugas LAB 1**

Nama  : Gilang Purnomo

NIM   : 2341720042

Absen : 14

## Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`

In [ ]:
# Load Data
import pandas as pd

df = pd.read_csv('spam.csv', encoding='latin-1')
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


In [ ]:
# Preprocessing
df = df[['v1', 'v2']]
df.columns = ['label', 'text']

df['label'] = df['label'].map({'spam': 1, 'ham': 0})

df.head()

label                                               text
0      0  Go until jurong point, crazy.. Available only ...
1      0                      Ok lar... Joking wif u oni...
2      1  Free entry in 2 a wkly comp to win FA Cup fina...
3      0  U dun say so early hor... U c already then say...
4      0  Nah I don't think he goes to usf, he lives aro...

2. Fitur `CountVectorizer` dengan mengaktifkan **stop_words**

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split

X = df['text']
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# CountVectorizer dengan stop words bahasa Inggris
vectorizer_count = CountVectorizer(stop_words='english')

# Transformasi teks ke dalam bentuk fitur numerik
X_train_count = vectorizer_count.fit_transform(X_train)
X_test_count = vectorizer_count.transform(X_test)

3. Evaluasi hasilnya

In [ ]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

mnb_count = MultinomialNB()

mnb_count.fit(X_train_count, y_train)

y_pred_count = mnb_count.predict(X_test_count)

print("Akurasi:", accuracy_score(y_test, y_pred_count))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_count))
print("Laporan Klasifikasi:\n", classification_report(y_test, y_pred_count))

Akurasi: 0.9838565022421525
Confusion Matrix:
 [[960   6]
 [ 12 137]]
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       0.99      0.99      0.99       966
           1       0.96      0.92      0.94       149

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



## Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data `spam.csv`

2. Fitur `TF-IDF` dengan mengaktifkan **stop_words**

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# init TF-IDF
vectorizer_tfidf = TfidfVectorizer(stop_words='english')

# transform teks ke bentuk fitur TF-IDF
X_train_tfidf = vectorizer_tfidf.fit_transform(X_train)
X_test_tfidf = vectorizer_tfidf.transform(X_test)

3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

In [ ]:
mnb_tfidf = MultinomialNB()

mnb_tfidf.fit(X_train_tfidf, y_train)

y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)

print("Akurasi:", accuracy_score(y_test, y_pred_tfidf))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_tfidf))
print("Laporan Klasifikasi:\n", classification_report(y_test, y_pred_tfidf))

Akurasi: 0.968609865470852
Confusion Matrix:
 [[966   0]
 [ 35 114]]
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       0.97      1.00      0.98       966
           1       1.00      0.77      0.87       149

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.97      1115



In [ ]:
accuracy_count = accuracy_score(y_test, y_pred_count)
accuracy_tfidf = accuracy_score(y_test, y_pred_tfidf)

print(f"Akurasi CountVectorizer: {accuracy_count:.4f}")
print(f"Akurasi TF-IDF: {accuracy_tfidf:.4f}")
print(f"Selisih Akurasi: {abs(accuracy_count - accuracy_tfidf):.4f}")

Akurasi CountVectorizer: 0.9839
Akurasi TF-IDF: 0.9686
Selisih Akurasi: 0.0152


4. Berikan kesimpulan fitur mana yang terbaik pada kasus data `spam.csv`

Dari hasil evaluasi, CountVectorizer menghasilkan akurasi tertinggi yaitu 98.4%, lebih baik dibandingkan TF-IDF (96.9%). Dari akurasi tersebut menunjukkan bahwa dataset spam.csv, frekuensi kemunculan kata (tanpa pembobotan TF-IDF) lebih efektif dalam membedakan pesan spam dan ham. Jadi, model terbaik di percobaan ini adalah MultinomialNB dengan fitur CountVectorizer.